source/common/upstream/health_discovery_service.cc

```diff
@@ -165,69 +164,141 @@ void HdsDelegate::onReceiveInitialMetadata(Http::ResponseHeaderMapPtr&& metadata
+                                   const envoy::config::cluster::v3::Cluster& cluster_config) {
+  cluster->update(admin_, cluster_config, info_factory_, cm_, local_info_, dispatcher_, random_,
+                  singleton_manager_, tls_, validation_visitor_, api_, access_log_manager_,
+                  runtime_);
+}
+
+HdsClusterPtr
+HdsDelegate::createHdsCluster(const envoy::config::cluster::v3::Cluster& cluster_config) {
+  static const envoy::config::core::v3::BindConfig bind_config;
+
+  // Create HdsCluster.
+  auto new_cluster = std::make_shared<HdsCluster>(
+      admin_, runtime_, std::move(cluster_config), bind_config, store_stats_, ssl_context_manager_,
+      false, info_factory_, cm_, local_info_, dispatcher_, random_, singleton_manager_, tls_,
+      validation_visitor_, api_);
+
+  // Begin HCs in the background.
+  new_cluster->initialize([] {});
+  new_cluster->initHealthchecks(access_log_manager_, runtime_, random_, dispatcher_, api_);
+
+  return new_cluster;
+}
+
 void HdsDelegate::processMessage(
     std::unique_ptr<envoy::service::health::v3::HealthCheckSpecifier>&& message) {
   ENVOY_LOG(debug, "New health check response message {} ", message->DebugString());
   ASSERT(message);
+  std::vector<HdsClusterPtr> hds_clusters;
+  // Maps to replace the current member variable versions.
+  absl::flat_hash_map<uint64_t, HdsClusterPtr> new_hds_clusters_hash_map;
+  absl::flat_hash_map<std::string, HdsClusterPtr> new_hds_clusters_name_map;
 
   for (const auto& cluster_health_check : message->cluster_health_checks()) {
-    // Create HdsCluster config
-    static const envoy::config::core::v3::BindConfig bind_config;
-    envoy::config::cluster::v3::Cluster cluster_config;
-
-    cluster_config.set_name(cluster_health_check.cluster_name());
-    cluster_config.mutable_connect_timeout()->set_seconds(ClusterTimeoutSeconds);
-    cluster_config.mutable_per_connection_buffer_limit_bytes()->set_value(
-        ClusterConnectionBufferLimitBytes);
-
-    // Add endpoints to cluster
-    for (const auto& locality_endpoints : cluster_health_check.locality_endpoints()) {
-      // add endpoint group by locality to config
-      auto* endpoints = cluster_config.mutable_load_assignment()->add_endpoints();
-      // if this group contains locality information, save it.
-      if (locality_endpoints.has_locality()) {
-        endpoints->mutable_locality()->MergeFrom(locality_endpoints.locality());
-      }
-
-      // add all endpoints for this locality group to the config
-      for (const auto& endpoint : locality_endpoints.endpoints()) {
-        endpoints->add_lb_endpoints()->mutable_endpoint()->mutable_address()->MergeFrom(
-            endpoint.address());
+    const uint64_t cluster_config_hash = MessageUtil::hash(cluster_health_check);
+    HdsClusterPtr cluster_ptr;
+
+    // If this cluster with the exact configuration is already being tracked, skip it.
+    auto cluster_map_pair = hds_clusters_hash_map_.find(cluster_config_hash);
```

Comment:
Does this need to be a map? Could it just be a set?

Reply:
I know you had mentioned it earlier that maybe I should be doing name lookup first, and I agree with that. I think if we do name look-up, then if there is a name match we do a compare inside of updateCluster(...) with a hash to see if there was an update, that way we only hash Clusters that have name matches. Also this means retaining one less map, so I went ahead and removed it. The code now reflects this.